### PR TITLE
Add GacelaConfig::registerGenericListener()

### PR DESCRIPTION
**Pull request by Chemaclass**
## 📚 Description

We added the possibility to register a listener to a particular event https://github.com/gacela-project/gacela/pull/215 
It would be useful to be able to register a listener to all posible events.

## 🔖 Changes

- Added `registerGenericListener()` to be able to register a generic listener for all events.
```php
GacelaConfig::registerGenericListener(callable $listener): void
```

## 🧪 A visual example
```php
$renderEvent = static function (GacelaEventInterface $event): void {
    echo sprintf("%s - %s\n", \get_class($event), $event->toString());
};

Gacela::bootstrap(__DIR__, static function (GacelaConfig $config) use ($renderEvent): void {
    // All of these individual specific listeners
    $config->registerSpecificListener(ResolvedClassCachedEvent::class, $renderEvent);
    $config->registerSpecificListener(ResolvedClassCreatedEvent::class, $renderEvent);
    $config->registerSpecificListener(ResolvedClassTriedFromParentEvent::class, $renderEvent);
    $config->registerSpecificListener(ResolvedCreatedDefaultClassEvent::class, $renderEvent);
    $config->registerSpecificListener(ReadPhpConfigEvent::class, $renderEvent);

    // would be the same as registering this generic listener
    $config->registerGenericListener($renderEvent);
});
```

## 🙇🏼  Especial thanks 

Credits to @matthiasnoback. I got the inspiration for this implementation thanks to the "[Layers, ports, and adapters workshop](https://github.com/matthiasnoback/layers-ports-and-adapters-workshop)" I did with him last year.

### CODE REVIEW

File: src/Framework/Bootstrap/GacelaConfig.php [hunk 1]
```diff
@@ -228,13 +231,21 @@ public function disableEventListeners(): self
...
     /**
      * Register a listener when some event happens.
      *
      * @param class-string $event
      * @param callable(GacelaEventInterface):void $listener
      */
-    public function registerListener(string $event, callable $listener): void
+    public function registerSpecificListener(string $event, callable $listener): void
     {
         $this->listenersPerEvent[$event][] = $listener;
```
Comment (JesusValera) [inline src/Framework/Bootstrap/GacelaConfig.php:250]: What about renaming `listenersPerEvent` to `specificListeners`?

File: src/Framework/Bootstrap/GacelaConfig.php [hunk 2]
```diff
@@ -228,13 +231,21 @@ public function disableEventListeners(): self
         return $this;
     }
 
+    /**
+     * Register a generic listener when any event happens.
```
Comment (JesusValera) [inline src/Framework/Bootstrap/GacelaConfig.php:235]: Maybe a better description would be helpful (take a look at the review comment), I guess more people will have the same questions as I do 😃

File: src/Framework/Bootstrap/GacelaConfig.php [hunk 3]
```diff
@@ -253,6 +264,7 @@ public function registerListener(string $event, callable $listener): void
      *     project-namespaces: list<string>,
      *     config-key-values: array<string,mixed>,
      *     are-event-listeners-enabled: bool,
+     *     generic-listeners: list<callable>,
      *     listeners-per-event: array<class-string,list<callable>>,
```
Comment (JesusValera) [inline src/Framework/Bootstrap/GacelaConfig.php:268]: Same here with the name.

**Review by JesusValera — COMMENTED** [review]
Not sure if I understand properly.
Are you forced to use the `GacelaEventInterface` in order to capture any other class that implements `GacelaEventInterface`? Can you use a specific event (eg: `ResolvedClassCachedEvent` class) instead of the `GacelaEventInterface`? In this scenario, what happens?

**Comment by Chemaclass** [comment]
> Not sure if I understand properly. Should you use the `GacelaEventInterface` in order to capture any other class that implements `GacelaEventInterface`?
> 
> Can you use a specific event (eg: `ResolvedClassCachedEvent` class) instead of the `GacelaEventInterface`? In this scenario, what happens?

Yes, you must use `GacelaEventInterface` on the event that you received in the generic listener, because that's the global minimum "contract level" of all gacela events. Something more generic would ended up in an error on runtime.

If you would be interested in a particular event within a generic listener you would have to use a casting to another concrete.

File: src/Framework/Bootstrap/GacelaConfig.php (hunk 2 above)
Comment (Chemaclass) [reply src/Framework/Bootstrap/GacelaConfig.php:235]: I added some more doc lines below.

**Review by JesusValera — APPROVED** [review]
I see, great!

**Pull request merged**